File: Tree/binary_search_tree/c++/binary_search_tree.cpp
#include <iostream>
using namespace std;


/*
This represents the structure of the node of the Binary Search Tree.
*/


struct node
{
    node * left;
    node * right;
    int data;
};


/*
This function adds new data into binary search tree (BST). 
It takes root of the tree as parameter along with the new data to be inserted.
*/


node * add(node * root,int val)
{
    if(root == NULL)
    {
        node * newNode = new node;
        newNode->data = val;
        newNode->left=NULL;
        newNode->right=NULL;
        return newNode;
    }
    if(val > root->data)
        root->right=add(root->right,val);
    else
        root->left=add(root->left,val);
    return root;
}


/*
This function prints the inorder traversal of the BST thus printing
a list of entered values in ascending order.It takes the root node as a parameter.
*/


void print_inorder(node *root)
{
    if(root==NULL)
    return;
    print_inorder(root->left);
    cout<<root->data<<" ";
    print_inorder(root->right);
    
}

int main()
{
    node * root =NULL;
    
    //  Testing insertion
    root=add(root,12);
    root=add(root,30);
    root=add(root,2);
    root=add(root,15);
    root=add(root,6);
    root=add(root,54);
    root=add(root,27);
    
    //Printing result
    print_inorder(root);
    return 0;
}

File: binary search tree/binary_search_tree.py
class Node:
    def __init__(self, data):
        self.left = None
        self.right = None
        self.data = data

class BinaryTree:
    def __init__(self):
        self.root = None

    def insert(self, data):
        if self.root is None:
            self.root = Node(data)
        else:
            self._insert(data, self.root)

    def _insert(self, data, node):
        if data < node.data:
            if node.left is None:
                node.left = Node(data)
            else:
                self._insert(data, node.left)
        else:
            if node.right is None:
                node.right = Node(data)
            else:
                self._insert(data, node.right)

    def search(self, data):
        if self.root is not None:
            return self._search(data, self.root)
        else:
            return None

    def _search(self, data, node):
        if data == node.data:
            return node
        elif data < node.data and node.left is not None:
            return self._search(data, node.left)
        elif data > node.data and node.right is not None:
            return self._search(data, node.right)

    def traverse_in_order(self):
        if self.root is not None:
            self._traverse_in_order(self.root)

    def _traverse_in_order(self, node):
        if node is not None:
            self._traverse_in_order(node.left)
            print(node.data)
            self._traverse_in_order(node.right)
